framework/Runners/ExternalRunner.py

```diff
@@ -1,274 +0,0 @@
-
-    if os.path.exists(executable):
-      executableFile = open(executable, "r")
-
-      firstTwoChars = executableFile.read(2)
-
-      if firstTwoChars == "#!":
-        realExecutable = shlex.split(executableFile.readline())
-        self.raiseAMessage("reading #! to find executable:" + repr(realExecutable))
-        # The below code should work, and would be better than findMsys,
-        # but it doesn't work.
-        # winExecutable = subprocess.check_output(['cygpath','-w',realExecutable[0]],shell=True).rstrip()
-        # print("winExecutable",winExecutable)
-        # realExecutable[0] = winExecutable
-        def findMsys():
-          """
-            Function to try and figure out where the MSYS64 is.
-            @ In, None
-            @ Out, dir, String, If not None, the directory where msys is.
-          """
-          dir = os.getcwd()
-          head, tail = os.path.split(dir)
-          while True:
-            if tail.lower().startswith("msys"):
-              return dir
-            dir = head
-            head, tail = os.path.split(dir)
-          return None
-        msysDir = findMsys()
-        if msysDir is not None:
-          beginExecutable = realExecutable[0]
-          if beginExecutable.startswith("/"):
-            beginExecutable = beginExecutable.lstrip("/")
-          winExecutable = os.path.join(msysDir, beginExecutable)
-          self.raiseAMessage("winExecutable" + winExecutable)
-          realExecutable[0] = winExecutable
-        else:
-          self.raiseAWarning("Could not find msys in "+os.getcwd())
-        commandSplit = realExecutable + [executable] + commandSplit[1:]
-        return commandSplit
-    return origCommand
-
-
-  def start(self):
-    """
-      Function to run the driven code
-      @ In, None
-      @ Out, None
-    """
-    oldDir = os.getcwd()
-    os.chdir(self.workingDir)
-    localenv = dict(os.environ)
-    outFile = open(self.output,'w', self.bufferSize)
-    if platform.system() == 'Windows':
-      self.command = self.expandForWindows(self.command)
-      self.raiseAMessage("modified command to" + repr(self.command))
-    self.__process = utils.pickleSafeSubprocessPopen(self.command,shell=True,stdout=outFile,stderr=outFile,cwd=self.workingDir,env=localenv)
-    os.chdir(oldDir)
-    self.started = True
-
-  def kill(self):
-    """
-      Function to kill the subprocess of the driven code
-      @ In, None
-      @ Out, None
-    """
-    self.raiseAWarning("Terminating "+self.__process.pid+' '+self.command)
-    self.__process.terminate()
-
-  def getWorkingDir(self):
-    """
-      Function to get the working directory path
-      @ In, None
-      @ Out, workingDir, string, working directory
-    """
-    return self.workingDir
-
-  def getOutputFilename(self):
-    """
-      Function to get the output filenames
-      @ In, None
-      @ Out, self.output, string, output filename root
-    """
-    return os.path.join(self.workingDir,self.output)
```

Comment:
We could have the Code create an ExternalRunner, but it started to feel like the ExternalRunner was just an added layer to the Code model. The ExternalRunner and the InternalRunner in my mind interacted very differently in the our code and so I wasn't convinced it was the right hierarchy to take. I admit I could be wrong on this.

Only the Code used the ExternalRunner.

As another point, the JobHandler did a lot of extra checking to see if it was dealing with an ExternalRunner. This didn't feel clean to me.  In an object-oriented approach, the JobHandler wouldn't care what kind of Runner it was dealing with it and it would allow dynamic typing to ensure that each Runner was processed correctly.